libcrux-ml-dsa/src/samplex4.rs

```diff
@@ -1,629 +1,147 @@
-            seed: [u8; 34],
-        ) -> [[PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A]; ROWS_IN_A] {
-            unsafe { matrix_A_avx2(seed) }
+        fn matrix_flat<SIMDUnit: Operations>(
+            columns: usize,
+            seed: &[u8],
+            matrix: &mut [PolynomialRingElement<SIMDUnit>],
+        ) {
+            #[cfg_attr(not(hax), target_feature(enable = "avx2"))]
+            #[allow(unsafe_code)]
+            unsafe fn inner<SIMDUnit: Operations>(
+                columns: usize,
+                seed: &[u8],
+                matrix: &mut [PolynomialRingElement<SIMDUnit>],
+            ) {
+                matrix_flat::<SIMDUnit, crate::hash_functions::simd256::Shake128x4>(
+                    columns, seed, matrix,
+                )
+            }
+            unsafe { inner(columns, seed, matrix) };
         }
     }
-
-    #[cfg_attr(not(hax), target_feature(enable = "avx2"))]
-    #[allow(unsafe_code)]
-    #[allow(non_snake_case)]
-    pub(crate) unsafe fn matrix_A_avx2<
-        SIMDUnit: Operations,
-        const ROWS_IN_A: usize,
-        const COLUMNS_IN_A: usize,
-    >(
-        seed: [u8; 34],
-    ) -> [[PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A]; ROWS_IN_A] {
-        match (ROWS_IN_A as u8, COLUMNS_IN_A as u8) {
-            #[cfg(feature = "mldsa44")]
-            (4, 4) => matrix_A_4_by_4::<
-                SIMDUnit,
-                crate::hash_functions::simd256::Shake128x4,
-                ROWS_IN_A,
-                COLUMNS_IN_A,
-            >(seed),
-            #[cfg(feature = "mldsa65")]
-            (6, 5) => matrix_A_6_by_5::<
-                SIMDUnit,
-                crate::hash_functions::simd256::Shake128x4,
-                ROWS_IN_A,
-                COLUMNS_IN_A,
-            >(seed),
-            #[cfg(feature = "mldsa87")]
-            (8, 7) => matrix_A_8_by_7::<
-                SIMDUnit,
-                crate::hash_functions::simd256::Shake128x4,
-                ROWS_IN_A,
-                COLUMNS_IN_A,
-            >(seed),
-            _ => unreachable!(),
-        }
-    }
-}
-
-#[allow(non_snake_case)]
-pub(crate) fn matrix_A_generic<
-    SIMDUnit: Operations,
-    Shake128: shake128::XofX4,
-    const ROWS_IN_A: usize,
-    const COLUMNS_IN_A: usize,
->(
-    seed: [u8; 34],
-) -> [[PolynomialRingElement<SIMDUnit>; COLUMNS_IN_A]; ROWS_IN_A] {
-    match (ROWS_IN_A as u8, COLUMNS_IN_A as u8) {
-        #[cfg(feature = "mldsa44")]
-        (4, 4) => matrix_A_4_by_4::<SIMDUnit, Shake128, ROWS_IN_A, COLUMNS_IN_A>(seed),
-        #[cfg(feature = "mldsa65")]
-        (6, 5) => matrix_A_6_by_5::<SIMDUnit, Shake128, ROWS_IN_A, COLUMNS_IN_A>(seed),
-        #[cfg(feature = "mldsa87")]
-        (8, 7) => matrix_A_8_by_7::<SIMDUnit, Shake128, ROWS_IN_A, COLUMNS_IN_A>(seed),
-        _ => unreachable!(),
-    }
 }
 
-#[cfg(feature = "mldsa44")]
+// Not inling this causes a 10x slow-down
```

Comment:
Would be great to open an issue so we understand why inlining is so crucial here, or if it is well understood, then some documentation on what is going on, so that we don't forget and make a mistake with inlining later.